modules/core/src/convert.hpp

```diff
@@ -0,0 +1,12 @@
+
+namespace cv
+{
+float convertFp16SW(short fp16);
+short convertFp16SW(float fp32);
```

Comment:
I mean, try to move implementations of these functions into header file under anonymous namespace.
So compilation unit with SSE* generated code will have one version of these functions, and AVX*(FP16) compilation unit will have own version too.
Anonymous namespace is needed to eliminate linker error about duplicate identifiers or unwanted code "replacement".

```
namespace {
#if !CV_FP16_TYPE
// const numbers for floating points format
const unsigned int kShiftSignificand    = 13;
const unsigned int kMaskFp16Significand = 0x3ff;
const unsigned int kBiasFp16Exponent    = 15;
const unsigned int kBiasFp32Exponent    = 127;
#endif
  
#if CV_FP16_TYPE
float convertFp16SW(short fp16)
{
    // Fp16 -> Fp32
    Cv16suf a;
    a.i = fp16;
    return (float)a.h;
}
#else
float convertFp16SW(short fp16)
{
    ...
}
#endif

#if CV_FP16_TYPE
short convertFp16SW(float fp32)
{
    // Fp32 -> Fp16
    Cv16suf a;
    a.h = (__fp16)fp32;
    return a.i;
}
#else
short convertFp16SW(float fp32)
{
    ...
}
#endif

} // namespace
```

Loops for tails should be left "as is" in optimized version too.